File: golang/2092.find-all-people-with-secret.go
func dfs(start int, adj map[int][]int, vis map[int]bool, resMap map[int]bool) {
    if vis[start] {
        return
    }

    vis[start] = true
    resMap[start] = true

    for _, child := range adj[start] {
        dfs(child, adj, vis, resMap)
    }
}

func findAllPeople(n int, meetings [][]int, firstPerson int) []int {

    secretsMap := map[int]bool{0: true, firstPerson: true}

    timeMap := make(map[int]map[int][]int)

    for _, meeting := range meetings {
        src, dst, time := meeting[0], meeting[1], meeting[2]
        if timeMap[time] == nil {
            timeMap[time] = make(map[int][]int)
        }

        timeMap[time][src] = append(timeMap[time][src], dst)
        timeMap[time][dst] = append(timeMap[time][dst], src)
    }

    timeList := make([]int, 0, len(timeMap))
    for time, _ := range timeMap {
        timeList = append(timeList, time)
    }

    sort.Ints(timeList)

    for _, time := range timeList {
        visited := make(map[int]bool)

        for src, _ := range timeMap[time] {
            if secretsMap[src] {
                dfs(src, timeMap[time], visited, secretsMap)
            }
        }
    }

    secrets := make([]int, 0, len(secretsMap))
    for src, _ := range secretsMap {
        secrets = append(secrets, src)
    }

    return secrets
}

File: cpp/2092.find-all-people-with-secret.cpp
class Solution {
private:
    void dfs(int start, const map<int, vector<int>>& adj, set<int>& vis, set<int>& res) {
        if (vis.find(start) != vis.end()) {
            return;
        }

        vis.insert(start);
        res.insert(start);

        for (auto child : adj.at(start)) {
            dfs(child, adj, vis, res);
        }
    }

public:
    vector<int> findAllPeople(int n, vector<vector<int>>& meetings, int firstPerson) {
        set<int> secrets({0, firstPerson});

        map<int, map<int, vector<int>>> timeMap;
        for (const auto& meeting : meetings) {
            timeMap[meeting[2]][meeting[0]].push_back(meeting[1]);
            timeMap[meeting[2]][meeting[1]].push_back(meeting[0]);
        }

        vector<int> timeList;
        timeList.reserve(timeMap.size());
        for (const auto& [k, _] : timeMap) {
            timeList.push_back(k);
        }
        std::sort(timeList.begin(), timeList.end());

        for (auto time : timeList) {
            set<int> visited;

            for (const auto& [src, _] : timeMap[time]) {
                if (secrets.find(src) != secrets.end()) {
                    dfs(src, timeMap[time], visited, secrets);
                }
            }
        }

        vector<int> res;
        res.assign(secrets.begin(), secrets.end());
        return res;
    }
};